### ISSUE TITLE
Link reported as broken after two hops in a 301 chain

### ISSUE THREAD

**Issue by frederickjh** [michaeltelford/broken_link_finder#1]
Hi!
The CMS I am using will first redirect (301) to the language version of the page if the language prefix has not be added (ie. de for German). If the CMS now has an other redirect (301) for the url with the language prefix. This creates a chain of two 301. The next hop hits the page and returns a http status code of 200.

broken_link_finder reports the first link as broken in its output report.

The Googlebot, in the [source](http://www.redirect-checker.org/how-many-redirects.php)[(s)](https://chrisberkley.com/blog/what-are-redirect-chains-seo/) I have found, will follow up to four chained 301s before giving up. Chrome, Firefox, and Edge will follow up to five chained 301s before returning the error message:  "Too many redirects. Can't load site."

Now granted 301 chains are not good for a lot of reasons but I would hardly consider a chain of two 301s a broken link.

This feature request is to add a sane default of three chained 301 as acceptable and produce no output. Four or more chained 301s could give a warning, as the page may not be indexed or displayed.

The other option would be to make the number of chained 301 before a link is reported as "broken" (or possibly another term to better identify the issue) a command line option that the user can change at run time.

Thanks for considering this feature request!
Frederick

**Comment by frederickjh** [comment]
I changed the site so that there is only one 301 redirect before the 200 success and broken_link_finder still reports the link as broken. Still not sure why.

Here are the headers for the 301 link and the 200 using `curl`.

```http
$ curl -I https://example.com/de/page1
HTTP/1.1 301 Moved Permanently
Date: Tue, 09 Jul 2019 18:44:44 GMT
Server: Apache
X-UA-Compatible: IE=edge
Content-language: de
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Generator: Drupal 8 (https://www.drupal.org)
X-Drupal-Cache: HIT
Location: /de/folder/page2#anchor-on-page2
Content-Type: text/html; charset=UTF-8

$ curl -I https://example.com/de/folder/page2#anchor-on-page2
HTTP/1.1 200 OK
Date: Tue, 09 Jul 2019 18:46:40 GMT
Server: Apache
Cache-Control: max-age=900, public
X-Drupal-Dynamic-Cache: UNCACHEABLE
Link: </de/folder/page2>; rel="canonical", </de/node/70>; rel="shortlink", </de/folder/page2>; rel="revision"
X-UA-Compatible: IE=edge
Content-language: de
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
Vary: Cookie
X-Generator: Drupal 8 (https://www.drupal.org)
X-Drupal-Cache: HIT
Last-Modified: Tue, 09 Jul 2019 18:11:03 GMT
ETag: "1562695863"
Content-Type: text/html; charset=UTF-8
```

Maybe this will help find the issue.

**Comment by michaeltelford** [comment]
Hi Frederick,

Thanks for raising the issue. You’re correct, redirects should not appear as broken links. 

As of broken_link_finder 0.2.1 redirects are followed; up to 5 hops before the link is declared broken. 

Can you firstly check what version of broken_link_finder you’re using? And check if the issue is still present with the latest version of the gem (0.4.0) please?

Hopefully this solves the issue. If not, could you email me a link to your site in production to help debug the issue? My email is:

michael.telford@live.com

Thanks,
Michael

**Comment by frederickjh** [comment]
I am new to Ruby and could not find a version command to output the version of the gem. I checked the `Gemfile.lock` and it is showing:

```
PATH
  remote: .
  specs:
    broken_link_finder (0.4.0)
      thor (= 0.20.3)
      thread (= 0.2)
      wgit (= 0.0.9)

GEM
  remote: https://rubygems.org/
  specs:
    addressable (2.6.0)
``` 
`<clip>`

So it looks to me that it is version 0.4.0, which is the most recent version. I will send you an email with the site and the link that is being reported as broken.

I have to wonder if it may have something to do with the fact that the **Location:** returned from the 301 contains an anchor.

**Comment by michaeltelford** [comment]
Yep, you’re running version 0.4.0 which supports redirects so it seems a valid issue. 

When I receive your email with the failing links I’ll investigate it. 

Thanks,
Michael

**Comment by michaeltelford** [comment]
The issue was that relative `Location` paths (redirects) weren't being supported, only absolute paths/urls. This has now been resolved in broken_link_finder version `0.4.1` (available now on https://rubygems.org).

@frederickjh since you're using bundler, you can get the latest release by typing:

`bundle update broken_link_finder`

Once updated, can you please retry your scenario and report back?

Thanks,
Michael

**Comment by frederickjh** [comment]
I have updated to version 0.4.1 and now the 301s are handled correctly and not reported as broken links.
